File: LinkedList/02_DoublyLinkedList.cpp
#include<iostream>
using namespace std;

class Node{
    public:
    int data;
    Node* next;
    Node* prev;
    Node(int data){
        this->data = data;
        this->next = NULL;
        this->prev = NULL;
    }
};

void insertAtHead(Node* &head, int data){
    Node* newNode = new Node(data);
    Node* temp = head;
    newNode->next = head;
    head->prev = newNode;
    head = newNode;
}

void displayLinkedList(Node* &head){
    cout<<"Linked list is: "<<endl;
    Node* temp = head;
    while(temp!=NULL){
        cout<<temp->data<<" ";
        temp=temp->next;
    }
    cout<<endl;
}

int main(){
    Node* node1 = new Node(10);
    Node* head = node1;
    insertAtHead(head, 20);
    insertAtHead(head, 30);
    displayLinkedList(head);
    return 0;
}

File: Python/LinkedList/02_DoublyLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
    

class DoublyLinkedList:
    def __init__(self):
        self.head = None
    
    def insertAtHead(self, data):
        newNode = Node(data)
        newNode.next = self.head
        self.head = newNode
    
    def insertAtTail(self, data):
        newNode = Node(data)
        currentNode = self.head
        while(currentNode.next):
            currentNode = currentNode.next
        currentNode.next = newNode
    
    def displayDoublyLinkedList(self):
        print("Linked list: ")
        currentNode = self.head
        while(currentNode):
            print(currentNode.data)
            currentNode = currentNode.next


list = DoublyLinkedList()
list.insertAtHead(10)
list.insertAtHead(20)
list.insertAtTail(30)
list.insertAtHead(50)
list.displayDoublyLinkedList()